### DISCUSSION TITLE
[NativeAOT] How do I receive a callback on library load in Linux?

### DISCUSSION THREAD

**Discussion by slxdy** [dotnet/runtime#104529] [Q&A] [answered]
I'd like to have a DllMain, but on Linux.

Edit: I know the dangers of using DllMain in managed, but I understand what I'm doing. I'm already hooking into another thread, which allows me to use the System API either way.

**Comment by colejohnson66** [comment]
You should never do DllMain in managed code

**Reply by slxdy** [reply]
There is a reason for why it exists. I don't need my code to be 'safe' either way

**Reply by colejohnson66** [reply]
Your code will not work if you do `DllMain`

**Reply by slxdy** [reply]
Right, right... Because I totally haven't used it in 10 different projects. Read the edit in my post

**Reply by slxdy** [reply]
I'm sorry for my negative reply, but the one thing I can't stand in this community is everyone being against unsafe code.
Guys, it's an official feature. If someone understands what they're doing, what's the point of unsafe code if every answer is just "don't"?
Next to that, `DllMain` in C# is also a manually implemented feature. There is a reason for it to exist. Saying that it won't work is just straight up a lie. Of course, it is true that a lot of stuff WILL NOT work, but if you know what you're doing, you can get around that.

**Reply by jkotas** [reply]
DllMain written in managed code won't work reliably. It is not a lie. You cannot get around that in your code. The root source of this reliability is in the interaction of the low-level GC suspension code in this repo and the Windows OS "loader lock". The problem is described in great detail in https://learn.microsoft.com/en-us/cpp/dotnet/initialization-of-mixed-assemblies . 

.NET Framework 1.0 allowed you to write DllMain in managed code (in managed C++). It turned out be major reliability problem. We spent a ton of time on diagnosing these issues in user apps and working with our users on refactoring their code to avoid running managed code in DllMain.

**Reply by slxdy** [reply]
The GC is running on a separate thread, right? So as long as I don't make win32 api calls I should he fine, no?

**Reply by colejohnson66** [reply]
Maybe, but there's no guarantee. That's why the team has refused to support such a scenario. For example, even if *you* don't use P/Invoke, practically anything involving I/O *will* call into Win32. And things that don't use Win32 have no guarantee that they won't use it in the future. And it's not just Win32; *anything* involving P/Invoke could break, even internal calls to the runtime. As @jkotas said, they are "not interested in building features that are not 100% reliable."

Sure, if all your `DllMain` does is some computations and variable setting, you'll probably be fine. But, again, there's no guarantee about that.

The best thing you'll ever get is [module initializers](https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/proposals/csharp-9.0/module-initializers); Those are as close to a `DllMain` as you'll get, but they are safe.

**Reply by jkotas** [reply]
For example: I assume that your managed DllMain is the first managed code executed. It means that the prolog of the DllMain is going to initialize the runtime, the GC, etc. It will end up calling all this code: https://github.com/dotnet/runtime/blob/57a2712a4d1d1e5f91ee7f6a1a1f745d547a15b7/src/coreclr/nativeaot/Bootstrap/main.cpp#L182-L203 . You can see there is a lot of Win32 calls, many of them are not safe for DllMain.

**Reply by slxdy** [reply]
But in that case, why does it never fail?

**Reply by jkotas** [reply]
The DllMain issues are effectively race conditions that repro intermittently and only when there is a bunch of other activity in the process.

If you have a simple (single threaded) test that just loads the library with DllMain, it will pretty much never fail.

**Comment by hez2010** [comment]
Check https://github.com/dotnet/runtime/issues/66546

**Reply by slxdy** [reply]
Well yes, but thats the thing, I need an alternative for Linux

**Reply by jkotas** [reply, marked as answer]
The portable version are C++ global static constructor (or their low-level platform-specific mechanism - DllMain on Windows, https://stackoverflow.com/questions/39540364/does-static-initialization-and-or-other-code-get-run-when-dlopening on Linux and other ELF-based platforms, ...). You can create C++ file with global static constructor that calls back into your method, compile it to .obj/o, and link this .obj/o to your app. It comes with the same disclaimer as DllMain. It is not a reliable solution.

**Reply by slxdy** [reply]
Thanks! Since DllMain is already supported, any plans on adding support for the Linux alternative? Maybe a managed attribute that could be interpreted by the NativeAOT compiler?

**Reply by jkotas** [reply]
We are not interested in building features that are not 100% reliable.

**Reply by CeSun** [reply]
> The portable version are C++ global static constructor (or their low-level platform-specific mechanism - DllMain on Windows, https://stackoverflow.com/questions/39540364/does-static-initialization-and-or-other-code-get-run-when-dlopening on Linux and other ELF-based platforms, ...). You can create C++ file with global static constructor that calls back into your method, compile it to .obj/o, and link this .obj/o to your app. It comes with the same disclaimer as DllMain. It is not a reliable solution.

Hello, how to include .o files in c# and link them with nativeaot with c#

**Reply by aadog** [reply]
> 可移植版本是 C++ 全局静态构造函数（或其低级平台特定机制 - Windows 上的 DllMain， Linux 和其他基于 ELF 的平台上的https://stackoverflow.com/questions/39540364/does-static-initialization-and-or-other-code-get-run-when-dlopening，...）。您可以创建带有全局静态构造函数的 C++ 文件，该文件会回调到您的方法中，将其编译为 .obj/o，然后将此 .obj/o 链接到您的应用。它带有与 DllMain 相同的免责声明。这不是一个可靠的解决方案。


After my testing, it does not work under android. It seems that the runtime is not started.

**Reply by CeSun** [reply]
> >The portable version are C++ global static constructor (or their low-level platform-specific mechanism - DllMain on Windows, https://stackoverflow.com/questions/39540364/does-static-initialization-and-or-other-code-get-run-when-dlopening on Linux and other ELF-based platforms, ...). You can create C++ file with global static constructor that calls back into your method, compile it to .obj/o, and link this .obj/o to your app. It comes with the same disclaimer as DllMain. It is not a reliable solution.
> 
> After my testing, it does not work under android. It seems that the runtime is not started.

Hello, I have also tested it. Nativeaot relies on the constructor of global static variables for runtime initialization. However, the callback when loading so will be executed before the constructor of global static variables, so it will crash.

**Reply by aadog** [reply]
What should we do, how to call initialization manually in c

**Reply by CeSun** [reply]
> What should we do, how to call initialization manually in c

Currently, there is no way

**Reply by aadog** [reply]
I used a conversion method and actively called it after using a so injection
